codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java

```diff
@@ -438,7 +441,7 @@ private void parseBodyAttributesStandard() {
                     if (read == '&') {
                         currentStatus = MultiPartStatus.DISPOSITION;
                         ampersandpos = currentpos - 1;
-                        setFinalBuffer(undecodedChunk.retainedSlice(firstpos, ampersandpos - firstpos));
+                        setFinalBuffer(undecodedChunk.copy(firstpos, ampersandpos - firstpos));
```

Comment:
hmm... retainedSlice(...) should be fine. This will do an extra memory copy which I think is not needed.

Reply:
As I wrote, there is a "old" bug, rarely catched, when the buffer `undecodedChunk` exceed the memory limit. When it reached this limit, there is a call to `undecodedChunk.discardReadBytes();`. I found this bug accidently, through my other new calls to the very same methods and errors in unit testing.

When this call occurs, the underlying buffers are changed, changing all HttpData values (of course, incorrectly).
To prevent this, I propose to copy it. I know this is higher in memory copy, but correct from final values, instead of current status.
Also, I changed this "high limit only" call to "each time it is relevant" (when new data can be added while there are enough space within the current buffer to fill it with this new data).

Another way would be to igonre totally this high limit and therefore not calling, nowhere, this `undecodedChunk.discardReadBytes();`, in order to keep final consistency.

WDYT?

Reply:
Of course, the second option (no more calls to `discardReadBytes()`) would lead to higher memory usage since the underlying buffer will be kept until the end.
Using "to disk" (or mixed way) management of HttpData can decrease this memory pressure, espcially for File upload (but not limited to), but as in mixed way, some HttpData could stay using the original undecodedChunk, this could lead in wrong data at the end, so my original proposition.

Any direction or idea ?

Reply:
I think we should only call `undecoded.discardReadBytes()` if `refCnt() == 1`.

Reply:
It could be a good idea !
I know this will be almost never true (due to `retainedSlice()` that will increment `refCnt()`), but at least, it will give a chance for it (as with `SeekAhead` algorithm that is almost never usable) and obviously limits the copy as much as possible.

I will git it a try. Thanks @normanmaurer for this idea !

Reply:
Done:

New benchmark results:
Benchmark                                                                           Mode  Cnt  Score   Error   Units
HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderBigAdvancedLevel   thrpt    6  2,248 ± 0,198  ops/ms
HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderBigDisabledLevel   thrpt    6  2,067 ± 1,219  ops/ms
HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderBigParanoidLevel   thrpt    6  1,109 ± 0,038  ops/ms
HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderBigSimpleLevel     thrpt    6  2,326 ± 0,314  ops/ms
HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderHighAdvancedLevel  thrpt    6  1,444 ± 0,226  ops/ms
HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderHighDisabledLevel  thrpt    6  1,462 ± 0,642  ops/ms
HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderHighParanoidLevel  thrpt    6  0,159 ± 0,003  ops/ms
HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderHighSimpleLevel    thrpt    6  1,522 ± 0,049  ops/ms

Almost the same as in previous code.